### ISSUE TITLE
[2.2.3] Failed to compile on CentOS7

### ISSUE THREAD

**Issue by pierrecdn** [acassen/keepalived#1965]
**Describe the bug**
At bumping to v2.2.3, I got a bunch of compilation warnings.

**To Reproduce**
```
$ ./configure --build=x86_64-redhat-linux-gnu --host=x86_64-redhat-linux-gnu --program-prefix= --disable-dependency-tracking --prefix=/usr --exec-prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --sysconfdir=/etc --datadir=/usr/share --includedir=/usr/include --libdir=/usr/lib64 --libexecdir=/usr/libexec --localstatedir=/var --sharedstatedir=/var/lib --mandir=/usr/share/man --infodir=/usr/share/info --disable-vrrp --enable-sha1 --enable-regex ':--enable-dump-threads}' --with-init=systemd
(...)
Keepalived configuration
------------------------
Keepalived version       : 2.2.3
Compiler                 : gcc gcc (GCC) 9.3.1 20200408 (Red Hat 9.3.1-2)
Preprocessor flags       : -D_GNU_SOURCE -I/usr/include/libnl3
Compiler flags           : -g -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic -Wall -Wextra -Wunused -Wstrict-prototypes -Wabsolute-value -Waddress-of-packed-member -Walloca -Walloc-zero -Warray-bounds=2 -Wattribute-alias=2 -Wbad-function-cast -Wc11-c2x-compat -Wcast-align -Wcast-qual -Wdate-time -Wdisabled-optimization -Wdouble-promotion -Wduplicated-branches -Wduplicated-cond -Wfloat-conversion -Wfloat-equal -Wformat-overflow -Wformat-security -Wformat-signedness -Wformat-truncation -Wframe-larger-than=5120 -Wimplicit-fallthrough=3 -Winit-self -Winline -Winvalid-pch -Wjump-misses-init -Wlogical-op -Wmissing-declarations -Wmissing-field-initializers -Wmissing-include-dirs -Wmissing-prototypes -Wnested-externs -Wnormalized -Wnull-dereference -Wold-style-definition -Woverlength-strings -Wpointer-arith -Wredundant-decls -Wshadow -Wshift-overflow=2 -Wstack-protector -Wstrict-overflow=4 -Wstringop-overflow=2 -Wstringop-truncation -Wsuggest-attribute=cold -Wsuggest-attribute=format -Wsuggest-attribute=malloc -Wsuggest-attribute=noreturn -Wsuggest-attribute=pure -Wsync-nand -Wtrampolines -Wundef -Wuninitialized -Wunknown-pragmas -Wunsafe-loop-optimizations -Wunsuffixed-float-constants -Wunused-const-variable=2 -Wvariadic-macros -Wwrite-strings -fPIE -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -O2
Linker flags             : -Wl,-z,relro -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -pie -Wl,-z,relro -Wl,-z,now
Extra Lib                : -lm -lcrypto -lssl -lnl-3 -lnl-genl-3 -lpcre2-8
Use IPVS Framework       : Yes
IPVS use libnl           : Yes
IPVS syncd attributes    : Yes
IPVS 64 bit stats        : Yes
HTTP_GET regex support   : Yes
fwmark socket support    : Yes
Use VRRP Framework       : No
Use BFD Framework        : No
SNMP vrrp support        : No
SNMP checker support     : No
SNMP RFCv2 support       : No
SNMP RFCv3 support       : No
DBUS support             : No
Use JSON output          : No
libnl version            : 3
Use IPv4 devconf         : No
Use iptables             : No
Use nftables             : No
init type                : systemd
systemd notify           : No
Strict config checks     : No
Build documentation      : No
Default runtime options  : -D
$ /usr/bin/make -j32 STRIP=/bin/true
(...)

  CC       check_file.o
BUILDSTDERR: ipvswrapper.c: In function 'format_srule':
BUILDSTDERR: ipvswrapper.c:129:30: error: expected ')' before 'PRIu32'
BUILDSTDERR:   129 |   return sprintf(buf, "Fwm %" PRIu32 "%s", srule->user.fwmark, srule->af == AF_INET6 ? " inet6" : "");
BUILDSTDERR:       |                              ^~~~~~~
BUILDSTDERR:       |                              )
BUILDSTDERR: ipvswrapper.c:129:28: warning: spurious trailing '%' in format [-Wformat=]
BUILDSTDERR:   129 |   return sprintf(buf, "Fwm %" PRIu32 "%s", srule->user.fwmark, srule->af == AF_INET6 ? " inet6" : "");
BUILDSTDERR:       |                            ^
BUILDSTDERR: ipvswrapper.c:131:2: warning: implicit declaration of function 'inet_ntop'; did you mean 'inet_ntop2'? [-Wimplicit-function-declaration]
BUILDSTDERR:   131 |  inet_ntop(srule->af, srule->af == AF_INET ? (const void *)&srule->nf_addr.ip : (const void *)&srule->nf_addr.in6, bufp, INET6_ADDRSTRLEN);
BUILDSTDERR:       |  ^~~~~~~~~
BUILDSTDERR:       |  inet_ntop2
BUILDSTDERR: ipvswrapper.c:131:2: warning: nested extern declaration of 'inet_ntop' [-Wnested-externs]
BUILDSTDERR: make[2]: *** [ipvswrapper.o] Error 1
BUILDSTDERR: make[2]: *** Waiting for unfinished jobs....
BUILDSTDERR: check_http.c: In function 'http_read_thread':
BUILDSTDERR: check_http.c:1498:4: warning: implicit declaration of function 'EVP_MD_CTX_free'; did you mean 'EVP_MD_CTX_create'? [-Wimplicit-function-declaration]
BUILDSTDERR:  1498 |    EVP_MD_CTX_free(req->context);
BUILDSTDERR:       |    ^~~~~~~~~~~~~~~
BUILDSTDERR:       |    EVP_MD_CTX_create
BUILDSTDERR: check_http.c:1498:4: warning: nested extern declaration of 'EVP_MD_CTX_free' [-Wnested-externs]
BUILDSTDERR: check_http.c: In function 'http_response_thread':
BUILDSTDERR: check_http.c:1558:18: warning: implicit declaration of function 'EVP_MD_CTX_new'; did you mean 'EVP_MD_CTX_init'? [-Wimplicit-function-declaration]
BUILDSTDERR:  1558 |   req->context = EVP_MD_CTX_new();
BUILDSTDERR:       |                  ^~~~~~~~~~~~~~
BUILDSTDERR:       |                  EVP_MD_CTX_init
BUILDSTDERR: check_http.c:1558:18: warning: nested extern declaration of 'EVP_MD_CTX_new' [-Wnested-externs]
BUILDSTDERR: check_http.c:1558:16: warning: assignment to 'EVP_MD_CTX *' {aka 'struct env_md_ctx_st *'} from 'int' makes pointer from integer without a cast [-Wint-conversion]
BUILDSTDERR:  1558 |   req->context = EVP_MD_CTX_new();
BUILDSTDERR:       |                ^
BUILDSTDERR: check_ssl.c: In function 'ssl_read_thread':
BUILDSTDERR: check_ssl.c:324:4: warning: implicit declaration of function 'EVP_MD_CTX_free'; did you mean 'EVP_MD_CTX_create'? [-Wimplicit-function-declaration]
BUILDSTDERR:   324 |    EVP_MD_CTX_free(req->context);
BUILDSTDERR:       |    ^~~~~~~~~~~~~~~
BUILDSTDERR:       |    EVP_MD_CTX_create
BUILDSTDERR: check_ssl.c:324:4: warning: nested extern declaration of 'EVP_MD_CTX_free' [-Wnested-externs] 
```

**Expected behavior**
Should compile, as with v2.2.2.

**Distro (please complete the following information):**
 - Name: CentOS
 - Version: 7.9
 - Architecture: x86_64

**Comment by pqarmitage** [comment]
@pierrecdn Many thanks for this report. This should be relatively straightforward to add configure/build time tests for whether the EVP_* functions are supported, and if not use the code that was removed in commit 75f856ed.

I will push an update later today.

**Comment by pqarmitage** [comment]
Commits 624e170 and fb26ebd should resolve the build errors. Commit e7f931b would also be worth including.

The inet_ntop and PRIu32 errors were due to a couple of header files not being included; for some reason the includes are not needed on Fedora 34.

The EVP_MD_CTX_new/free errors were due to the function names being changed from EVP_MD_CTX_create/destroy with OpenSSL release 1.1.0, and I had developed the code with release 1.1.1k.

**Comment by pierrecdn** [comment]
> I note you are using the configure option --enable-sha1, which only applied to the genhash build, and it is no longer supported since the genhash functionality has been incorporated into the keepalived executable, and keepalived only supports using MD5 hashes.

Oh yes, this is a leftover, I actually have a warning saying that this is deprecated.

> I am somewhat confused by the gcc version that is being reported. My CentOS 7 system has gcc version 4.8.5, and my RHEL8 system only has version 8.4.1, whereas you have version 9.3.1.

Yes, that's something custom we've made to be able to compile on a recent platform.

> If you are getting particular warnings that you are worried about then please post the details and I will have a look, however I don't have access to gcc v9.3.1.

I'll test with the commits you mentioned, many thanks!

**Comment by shmux8** [comment]
Hi folks,

Thank you for patching this, but I still couldn't compile laste version on CentOS 7 when I do `./configure --disable-lvs`:
```
  CCLD     keepalived
vrrp/libvrrp.a(vrrp_ipsecah.o): In function `hmac_md5':
/tmp/keepalived/keepalived/vrrp/vrrp_ipsecah.c:84: undefined reference to `EVP_MD_CTX_new'
/tmp/keepalived/keepalived/vrrp/vrrp_ipsecah.c:93: undefined reference to `EVP_MD_CTX_reset'
/tmp/keepalived/keepalived/vrrp/vrrp_ipsecah.c:99: undefined reference to `EVP_MD_CTX_free'
/tmp/keepalived/keepalived/vrrp/vrrp_ipsecah.c:52: undefined reference to `EVP_MD_CTX_new'
/tmp/keepalived/keepalived/vrrp/vrrp_ipsecah.c:62: undefined reference to `EVP_MD_CTX_free'
```

**Comment by pierrecdn** [comment]
@pqarmitage On my side it now complains about `'aclocal-1.16' is missing on your system`. 
I guess this is due to my automake package being in version 1.13. Except fb26ebd I don't see any change that could trigger such thing. Even with this one, I don't understand the hard constraint on recent automake... :thinking:

**Comment by pqarmitage** [comment]
Many apologies for keepalived still failing to build. I added the definitions of EVP_MD_CTX_{new,reset,free} in the wrong place in configure.ac; it fixed the problem if building without `--disable-lvs` but unfortunately I didn't test with `--disable-lvs`. Commit 3e5cffe now moves the definitions to the correct place in configure.ac.

Regarding complaining about aclocal-1.16 missing, I think that is because the patch fb26ebd updates configure.ac, and so configure needs regenerating, and some of the supporting files included in the tarball are for aclocal-1.16 but your system has aclocal-1.13.

Can you try the following sequence of commands and see if that sorts it out for you:
```
make distclean     # This may fail, but if so don't worry
./autogen.sh
./configure --disable-lvs
make
```

**Comment by pierrecdn** [comment]
OK got it, succeeded to build now, thanks!
